ion-hash/src/representation.rs

```diff
@@ -6,40 +6,50 @@
 //! instead. This implementation fills in that gap, and is focused on coverage
 //! and not speed.
 
-use crate::Markers;
-use digest::{FixedOutput, Update};
+use crate::{
+    mark_begin, mark_end, type_qualifier::type_qualifier_symbol,
+    update_type_qualifier_and_representation, Markers,
+};
+use digest::{FixedOutput, Output, Reset, Update};
 use ion_rs::{
     binary,
-    value::{AnyInt, Element},
+    result::{illegal_operation, IonResult},
+    value::{AnyInt, Element, Struct, SymbolToken},
     IonType,
 };
 
 // TODO: Finish ion-rust's binary writer and factor it such that the binary
 // representations can be written by the "raw" writer (ref. the Java
 // implementation).
-pub(crate) fn update_with_representation<E: Element + ?Sized, D: Update + FixedOutput + Default>(
-    elem: &E,
-    hasher: &mut D,
-) {
+pub(crate) fn update_with_representation<E, D>(elem: &E, hasher: &mut D) -> IonResult<()>
+where
+    E: Element + ?Sized,
+    D: Update + FixedOutput + Reset + Clone + Default,
```

Comment:
The `D: ...` sequence is just the subtraits of the `Digest` trait.

Reply:
Have we come full circle here and proposed [`Digest`](https://docs.rs/digest/0.9.0/digest/trait.Digest.html).  I realize it is slightly different than `Digest` (using the bounds on `Self`) rather than a parallel API, but the blanket implementation is the same, so by definition we've kind of gone that route...

Reply:
I think it's different because the current decomposition allows for impls of just the bits we need. Either way, I will clean this up in a future PR.